### PR TITLE
Fft improvements

### PR DESCRIPTION
**Pull request by mfoerste4**
This PR is a re-factorization of the cufft wrapper task in order to

* remove the 3D-FFT limit
* allow for data parallel execution if feasible
* extend cuftPlanCache for advanced layouts

NOTE: This is a new PR based on #700 which had trouble with the CI due to the branch name.

### CODE REVIEW

**Comment by mfoerste4** [comment]
@magnatelee, ready for another review round - the cufft plan cache (#729 ) improves the startup time of asynchronous cufft tasks significantly.

File: cunumeric/deferred.py [hunk 1]
```diff
@@ -1427,7 +1427,10 @@ def fft(
             for ax in axes:
                 task.add_scalar_arg(ax, ty.int64)
 
-            task.add_broadcast(input)
+            if input.ndim > len(set(axes)):
```
Comment (magnatelee) [inline cunumeric/deferred.py:1430]: I think the `axes` should be sanitized before we get to this function.

File: cunumeric/fft/fft.py [hunk 2]
```diff
@@ -103,7 +103,7 @@ def fft(
 
     Availability
     --------
-    Single GPU
+    Multiple GPUs
```
Comment (magnatelee) [inline cunumeric/fft/fft.py:106]: Same comment here. Let's make some notes on when we can parallelize FFTs.

File: src/cunumeric/cuda_help.h [hunk 3]
```diff
@@ -108,14 +108,41 @@ class cufftContext {
   std::vector<cufftXtCallbackType> callback_types_{};
 };
 
+struct cufftPlanParms {
```
Comment (magnatelee) [inline src/cunumeric/cuda_help.h:111]: a nitpick: "parms" seems a bit unusual abbreviation for "parameters". ("params" would be a typical name)

**Review by magnatelee — CHANGES_REQUESTED** [review]

**Pull request merged**